File: packages/ash-dotnet/Stores/MemoryStore.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;

namespace Ash.Stores;

/// <summary>
/// In-memory implementation of IContextStore.
///
/// Suitable for development and single-instance deployments.
/// For production with multiple instances, use Redis store.
/// </summary>
public class MemoryStore : IContextStore
{
    private readonly ConcurrentDictionary<string, AshContext> _contexts = new();

    /// <summary>
    /// Create a new context.
    /// </summary>
    public Task<AshContext> CreateAsync(
        string binding,
        long ttlMs,
        AshMode mode,
        Dictionary<string, object>? metadata)
    {
        var id = GenerateContextId();
        var nonce = mode == AshMode.Strict ? GenerateNonce() : null;

        var context = new AshContext
        {
            Id = id,
            Binding = binding,
            ExpiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + ttlMs,
            Mode = mode,
            Used = false,
            Nonce = nonce,
            Metadata = metadata
        };

        _contexts[id] = context;
        return Task.FromResult(context);
    }

    /// <summary>
    /// Get a context by ID.
    /// </summary>
    public Task<AshContext?> GetAsync(string id)
    {
        if (!_contexts.TryGetValue(id, out var context))
        {
            return Task.FromResult<AshContext?>(null);
        }

        if (context.IsExpired())
        {
            _contexts.TryRemove(id, out _);
            return Task.FromResult<AshContext?>(null);
        }

        return Task.FromResult<AshContext?>(context);
    }

    /// <summary>
    /// Consume a context.
    /// </summary>
    public Task<bool> ConsumeAsync(string id)
    {
        if (!_contexts.TryGetValue(id, out var context))
        {
            return Task.FromResult(false);
        }

        if (context.Used || context.IsExpired())
        {
            return Task.FromResult(false);
        }

        context.Used = true;
        return Task.FromResult(true);
    }

    /// <summary>
    /// Clean up expired contexts.
    /// </summary>
    public Task<int> CleanupAsync()
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var removed = 0;

        foreach (var kvp in _contexts)
        {
            if (now > kvp.Value.ExpiresAt)
            {
                if (_contexts.TryRemove(kvp.Key, out _))
                {
                    removed++;
                }
            }
        }

        return Task.FromResult(removed);
    }

    /// <summary>
    /// Get the number of active contexts.
    /// </summary>
    public int Count => _contexts.Count;

    /// <summary>
    /// Clear all contexts.
    /// </summary>
    public void Clear() => _contexts.Clear();

    private static string GenerateContextId()
    {
        var bytes = RandomNumberGenerator.GetBytes(16);
        return "ctx_" + Convert.ToHexString(bytes).ToLowerInvariant();
    }

    private static string GenerateNonce()
    {
        var bytes = RandomNumberGenerator.GetBytes(16);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }
}

File: packages/ash-php/src/Store/MemoryStore.php
<?php

declare(strict_types=1);

namespace Ash\Store;

use Ash\AshContext;
use Ash\AshMode;

/**
 * In-memory context store.
 *
 * Suitable for development and single-process deployments.
 * For production, use Redis or database store.
 */
final class MemoryStore implements ContextStoreInterface
{
    /** @var array<string, AshContext> */
    private array $contexts = [];

    /**
     * Create a new context.
     */
    public function create(
        string $binding,
        int $ttlMs,
        AshMode $mode,
        array $metadata = []
    ): AshContext {
        $id = 'ctx_' . bin2hex(random_bytes(16));
        $nonce = $mode === AshMode::Strict ? bin2hex(random_bytes(16)) : null;

        $context = new AshContext(
            id: $id,
            binding: $binding,
            expiresAt: $this->getTimestampMs() + $ttlMs,
            mode: $mode,
            used: false,
            nonce: $nonce,
            metadata: $metadata,
        );

        $this->contexts[$id] = $context;

        return $context;
    }

    /**
     * Get a context by ID.
     */
    public function get(string $id): ?AshContext
    {
        $context = $this->contexts[$id] ?? null;

        if ($context === null) {
            return null;
        }

        // Check expiration
        if ($context->isExpired()) {
            unset($this->contexts[$id]);
            return null;
        }

        return $context;
    }

    /**
     * Consume a context.
     */
    public function consume(string $id): bool
    {
        $context = $this->get($id);

        if ($context === null) {
            return false;
        }

        if ($context->used) {
            return false;
        }

        // Mark as used (create new instance since properties are readonly)
        $this->contexts[$id] = new AshContext(
            id: $context->id,
            binding: $context->binding,
            expiresAt: $context->expiresAt,
            mode: $context->mode,
            used: true,
            nonce: $context->nonce,
            metadata: $context->metadata,
        );

        return true;
    }

    /**
     * Clean up expired contexts.
     */
    public function cleanup(): int
    {
        $now = $this->getTimestampMs();
        $removed = 0;

        foreach ($this->contexts as $id => $context) {
            if ($now > $context->expiresAt) {
                unset($this->contexts[$id]);
                $removed++;
            }
        }

        return $removed;
    }

    /**
     * Get the number of active contexts.
     */
    public function count(): int
    {
        return count($this->contexts);
    }

    /**
     * Clear all contexts.
     */
    public function clear(): void
    {
        $this->contexts = [];
    }

    /**
     * Get current timestamp in milliseconds.
     */
    private function getTimestampMs(): int
    {
        return (int)(microtime(true) * 1000);
    }
}
